tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,199 @@
+	*asp.IdentStatement
+	StartLine int
+	EndLine int
+}
+
+func newAnalyzer() *Analyzer {
+	state := core.NewDefaultBuildState()
+	parser := parse.GetParserWithBuiltins(state)
+
+	a := &Analyzer{
+		parser:parser,
+	}
+	a.BuiltInsRules()
+
+	return a
+}
+
+// BuiltInsRules gets all the builtin functions and rules as a map, and store it in Analyzer.BuiltIns
+func (a *Analyzer) BuiltInsRules() {
+	statementMap := make(map[string]*RuleDef)
+
+	for _, statement := range a.parser.GetAllBuiltinStatements() {
+		// Saves FuncDefs into the statementMap if it's a None private rule
+		if statement.FuncDef != nil && !strings.HasPrefix(statement.FuncDef.Name, "_") {
+			statementMap[statement.FuncDef.Name] = newRuleDef(statement.FuncDef)
+		}
+	}
+	a.BuiltIns = statementMap
+}
+
+// IdentFromPos gets the Identifier given a lsp.Position
+func (a *Analyzer) IdentFromPos(uri lsp.DocumentURI, position lsp.Position, filecontent []string) (*Identifier, error) {
+	idents, err := a.IdentFromFile(uri, filecontent)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ident := range idents {
+		if position.Line >= ident.StartLine && position.Line <= ident.EndLine {
+			return ident, nil
+		}
+	}
+
+	return nil, nil
+}
+
+
+// IdentFromFile gets all the Identifiers from a given BUILD file
+// *reads complete files only*
+func (a *Analyzer) IdentFromFile(uri lsp.DocumentURI, filecontent []string) ([]*Identifier,  error) {
+	filepath, err := GetPathFromURL(uri, "file")
+	if err != nil {
+		return nil, err
+	}
+	bytecontent, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	stmt, err := a.parser.ParseData(bytecontent, filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var idents []*Identifier
+	for i, val := range stmt {
+		//fmt.Println(val.Ident.Action.Call)
+		if val.Ident != nil {
+			ident := &Identifier{
+				IdentStatement: val.Ident,
+				// -1 from asp.Statement.Pos.Line, as lsp position requires zero index
+				StartLine: val.Pos.Line - 1,
+			}
+			idents = append(idents, ident)
+
+			// fillin End for the previous iterated identifier
+			if i > 0 {
+				fmt.Println(ident)
+				idents[i - 1].EndLine = ident.StartLine - 1
+			}
+		}
+	}
+
+	// Finally, we fill in the last of the endline
+	idents[len(idents) - 1].EndLine = len(filecontent)
+
+	return idents, nil
+}
+
+func newRuleDef(funcDef *asp.FuncDef) *RuleDef {
+	ruleDef := &RuleDef{
+		FuncDef:funcDef,
+		ArgMap: make(map[string]*Argument),
+	}
+
+	var headerSlice []string
+	header := "def " + funcDef.Name + "("
+	line := header
+	argLen := len(funcDef.Arguments)
+	if argLen > 0 {
+		for i, arg := range funcDef.Arguments {
+			// Check if it a builtin type method, and reconstruct header if it is
+			if i == 0 && arg.Name == "self" {
+				line = arg.Type[0] + "." +funcDef.Name + "("
+			} else {
+				// Fill in the ArgMap
+				argString := getArgument(arg)
+				required := true
+				if strings.Contains(argString, "=") {
```

Comment:
yeah look at the Value, that will give you this more elegantly.

You could also write it inline, like 
```
required: arg.Value == nil,
```